Batteries/Data/List/Lemmas.lean

```diff
@@ -478,6 +478,20 @@ theorem length_eraseIdx : ∀ {l i}, i < length l → length (@eraseIdx α l i)
 
 @[simp] theorem length_tail (l : List α) : length (tail l) = length l - 1 := by cases l <;> rfl
 
+theorem succ_length_tail (l : List α) (h : 0 < length l) : (length (tail l)).succ = length l := by
+  cases l <;> simp at h ⊢
```

Comment:
I really wish `simp [*]` or `simp_all` could just prove this. There might be a missing simp somewhere in `Nat`.

Here is a different proof but not better than yours.
```lean
theorem succ_length_tail (l : List α) (h : 0 < length l) : (length (tail l)).succ = length l := by
  cases l; contradiction; simp
```

Reply:
I don't know about the missing `simp` stuff, but I do think your version shows the argument more clearly, so I adopted that one.

Reply:
There is definitely a `simp` issue here since this minor variant works:
```lean
theorem length_tail_add_one (l : List α) (h : 1 ≤ length l) : (length (tail l)) + 1 = length l := by simp [h]
```
I'll keep investigating...

In any case, we do prefer `add_one` over `succ` since `Nat` induction now uses `_+1` instead of `succ`. So `Nat.succ` is considered low-level and most end-user theorems should try to use `_+1` instead.

Reply:
Fixed the comments. I've also looked a bit closer at the simp issue and I've pushed some stuff that might address it. I don't know if this is something that belongs in batteries, but it does seem a bit detailed for the core repo?

Reply:
The `Nat` lemmas should go in a separate PR.

Another solution is to add `@[simp]` to `Nat.sub_one_add_one` and `Nat.sub_one_add_one_eq_of_pos` in Lean core.

Reply:
Okay, I pulled them out of this one:) I'm not great at overseeing the consequences of adding simp lemmas, do you have a preference for either solution?

Reply:
Conditional simp lemmas are a bit tricky. When multiple such lemmas trigger on the same expression, in this case `(a - 1) + 1` in this case, they each trigger a search for the requisite hypothesis. So, fewer is better and an additional conditional simp lemma with an unlikely hypothesis is not useful. From that point of view, my solution is worse since it adds two conditional simp lemmas but then again your solution triggers on a more general expression, namely `(a - b) + b`.

Maybe there's an even better solution we're not seeing. For example, simprocs may be useful in this situation but I've delayed learning how to properly use these, so I don't know if they could actually help. Bottom line: I don't know what the right thing to do is here.

(Just in case that wasn't clear: I only recommended splitting the `Nat` lemmas because focused PRs get merged faster, not because I thought your solution was worse!)

Reply:
Gotcha:) In that case I would like to propose merging this as-is and postponing the simp-related stuff to possible future PRs.